Code/Framework/AzCore/AzCore/IO/Streamer/RecentlyUsedIndex.inl

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+namespace AZ::IO
+{
+    template<typename T>
+    RecentlyUsedIndex<T>::RecentlyUsedIndex(T size)
+        : m_indices_previous(AZStd::make_unique<T[]>(size))
+        , m_indices_next(AZStd::make_unique<T[]>(size))
+        , m_size(size)
+        , m_front(0)
+        , m_back(size - 1)
+    {
+        AZ_Assert(size > 2, "At least 2 entries are needed in order to have a separate front and back.");
+        AZ_Assert(size < AZStd::numeric_limits<T>::max() - 1, "Size too large for Recently Used Index with the provided type.");
+
+        FlushAll();
+    }
+
+    template<typename T>
+    void RecentlyUsedIndex<T>::Touch(T index)
+    {
+        if (index != m_back)
+        {
+            AZ_Assert(index < m_size, "Index out of bounds while calling Touch in Recently Used Index.");
+
+            // Remove the index from its current spot and fix m_front if needed.
+            T previous = m_indices_previous[index];
+            T next = m_indices_next[index];
+            if (previous != InvalidIndex)
+            {
+                m_indices_next[previous] = next;
+            }
+            else
+            {
+                m_front = next; // This index was at the front, so the front needs to be reset.
+            }
+
+            if (next != InvalidIndex)
+            {
+                m_indices_previous[next] = previous;
+            }
+
+            // Reset indices.
+            m_indices_previous[index] = m_back;
+            m_indices_next[index] = InvalidIndex;
+
+            // Insert at the back.
+            m_indices_next[m_back] = index;
+            m_back = index;
+        }
+    }
+
+    template<typename T>
+    void RecentlyUsedIndex<T>::Flush(T index)
+    {
+        if (index != m_front)
+        {
+            AZ_Assert(index < m_size, "Index out of bounds while calling Flush in Recently Used Index.");
+
+            T previous = m_indices_previous[index];
+            T next = m_indices_next[index];
+            if (previous != InvalidIndex)
+            {
+                m_indices_next[previous] = next;
+            }
+
+            if (next != InvalidIndex)
+            {
+                m_indices_previous[next] = previous;
+            }
+            else
+            {
+                m_back = previous; // This index was at the front, so the front needs to be reset.
+            }
```

Comment:
This part of the logic is the same with Touch.
This can be candidate to refactor to some kind of private `Unlink` function that "unlinks" the supplied index from the list.

Reply:
There's actually a small difference because `m_front` or `m_back` doesn't need to be checked as it'll be reset regardless later in the call. Still that's easy enough to encapsulate with a template argument so moved them into their own function in the next update.